src/mono/mono/mini/mini-llvm.c

```diff
@@ -6195,13 +6205,19 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 			case LLVMArgFpStruct: {
 				LLVMTypeRef ret_type = LLVMGetReturnType (ctx->lmethod_type);
 				LLVMValueRef retval, elem;
-				gboolean is_simd = mini_class_is_simd (ctx->cfg, mono_class_from_mono_type_internal (sig->ret));
+
+				MonoClass *klass=  mono_class_from_mono_type_internal (sig->ret);
+				gboolean is_simd = mini_class_is_simd (ctx->cfg, klass);
 
 				if (is_simd) {
 					retval = LLVMConstNull(ret_type);
 
 					if (lhs) {
 						int len = LLVMGetVectorSize (LLVMTypeOf (lhs));
+
+						if (mono_class_value_size (klass, NULL) == 12) 
+							len--;
```

Comment:
Why are we decrementing `len` here?

Reply:
I encountered case that `lhs` would be V3 represented as V4(three elements + zero). And the `retval` would be V3. 

The code tried to extract the fourth value of `lhs` and insert it into `retval` which caused an error, as `retval` had only three elements

Reply:
I would add comment explaining this special case for easier comprehension.

Reply:
@jkurdek thank you for the explanation. 

Just for my understanding, so it basically means that in such cases `LLVMGetReturnType (ctx->lmethod_type)` and `LLVMTypeOf (lhs)` are returning a different type (as we end up having different vector sizes) ?

If we have these kind of workarounds across the mini-llvm code, I think it is ok to manually adjust the `len` however to be less cryptic, it might be better to have a comment explaining the special-casing (as @matouskozak suggested) and something like this instead:
```c
// TODO: Here add explanation why handling Vector3 requires special treatment
int len = mono_class_value_size (klass, NULL) == 12 ? 3 : LLVMGetVectorSize (LLVMTypeOf (lhs));
```

Reply:
Yup, `LLVMGetReturnType (ctx->lmethod_type)` is a vector of three elements, `whereas LLVMTypeOf (lhs)` is a vector of four elements (three + zero)

Reply:
Since 3 element vectors don't exist in the hardware, it might be easier to use 4 elements ones in return types etc. and only handle 3 element ones in loads/stores etc.